nautobot/extras/models/groups.py

```diff
@@ -431,3 +446,199 @@ def clean(self):
+
+    def generate_query_for_group(self, group):
+        fs = group.filterset_class(group.filter, group.get_queryset())
+
+        q = models.Q()
+
+        for field_name, value in fs.data.items():
+            filter_field = fs.filters[field_name]
+            q &= self.generate_query_for_filter(filter_field, value)
+
+        return q
+
+    def process_group_filters(self, group=None, qs=None):
+        # Start with this group's base queryset
+        base_qs = self.get_queryset()
+        if qs is None:
+            # Clone the base queryset so we can modify it.
+            qs = self.get_queryset()
+
+        if group is None:
+            group = self
+            logger.debug("Processing group %s...", group)
+
+        big_q = models.Q()
+
+        # Enumerate the filters. Recursing into any groups of groups.
+        for membership in group.dynamic_group_memberships.all():
+            group = membership.group
+            logger.debug("Processing group %s...", group)
+            child_memberships = group.dynamic_group_memberships.all()
+            if child_memberships.exists():
+                qs = self.process_group_filters(group=group, qs=qs)
+
+            operator = membership.operator
+
+            logger.debug("%s -> %s -> %s", group, group.filter, operator)
+
+            # The previous iteration that used set operations at the DB layer
+            # but limited by not being able to do further filtering.
+            """
+            fs = self.filterset_class(group.filter, self.get_queryset())
+            if operator == "union":
+                qs = qs.union(fs.qs)
+            elif operator == "difference":
+                qs = qs.difference(fs.qs)
+            elif operator == "intersection":
+                qs = qs.intersection(fs.qs)
+            """
+            next_set = self.generate_query_for_group(group)
+            # print(next_set)
+            if operator == "union":
+                big_q |= next_set
+                qs = qs | base_qs.filter(next_set)
+            elif operator == "difference":
+                big_q &= ~next_set
+                qs = qs.exclude(next_set)
+            elif operator == "intersection":
+                big_q &= next_set
+                qs = qs.filter(next_set)
+            # print(big_q)
+
+        # return big_q
+        return qs
+
+    def get_group_queryset(self):
+        return self.process_group_filters(group=self)
+
+    def get_descendants(self, group=None, descendants=None):
+        """Return the children of all child groups."""
+        if group is None:
+            group = self
+        if descendants is None:
+            descendants = []
+
+        for child_group in group.children.all():
+            print(f"Processing group {child_group}...")
+            # descendants.append(child_group.pk)
+            descendants.append(child_group)
+            if child_group.children.exists():
+                self.get_descendants(child_group, descendants)
```

Comment:
Took me a double-take to realize why you're not using the return value from this call, because you're passing `descendants` by reference and modifying it directly. Would it be more obvious/intuitive to instead do something like:

```python
if child_group.children.exists():
    descendants.extend(self.get_descendants(child_group))
```

Reply:
No because it needs to be recursive. Not sure if your suggestion would still satisfy that?

Reply:
And also, this will likely be replaced entirely by making `DynamicGroup` a tree model a la MPTT/tree-queries.

Reply:
I misread before, and you're right. Thanks.